### ISSUE TITLE
Latex not working in jupyter notebook v7.

### ISSUE THREAD

**Issue by pfebrer** [plotly/plotly.py#4336]
I recently upgraded to jupyter notebook v7 and I noticed that some plotly plots that previously rendered flawlessly now fail to render. I get the following output:

![Screenshot from 2023-08-22 12-27-21](https://github.com/plotly/plotly.py/assets/42074085/534753f0-449b-4a8a-a79e-1f8a27f22d57)

After some exploration I found out that it comes from an error using MathJax in `jupyterlab-plotly`. It tries to access `MathJax.config.startup.output`, but `MathJax.config.startup` is `undefined`.

It is very easy to reproduce this error:
- Create a new python environment (I tried with python 3.8).
- `pip install jupyter plotly`
- Open a jupyter notebook with `jupyter notebook`.
- Run a cell with the following code:

```python
from plotly.graph_objects import Figure

Figure().update_layout(title="$Title$")
```

**Comment by pfebrer** [comment]
Ok, I installed the `jupyterlab-plotly` extension with `jupyter labextension install plotlywidget` and it works, so I guess this will be fixed in the next python package release (?)

**Comment by pfebrer** [comment]
No, sorry it also doesn't work, the previous code snippet just literally shows "$Title$" as a title, but prepending the string with 'r' makes the issue show again:

```python
from plotly.graph_objects import Figure

Figure().update_layout(title=r"$Title$")
```

**Comment by zhimin-z** [comment]
Hi, I faced the same issue, any update? @pfebrer

**Comment by pfebrer** [comment]
No, I still need some solution for this 😅

**Comment by newville** [comment]
@zhimin-z @pfebrer I've been seeing this issue for a while too.  Is there any understanding of the cause and what might be done to fix it?

**Comment by zhimin-z** [comment]
Yeah, still there... @newville 
![image](https://github.com/plotly/plotly.py/assets/8592144/091a34b2-294e-45c7-be68-a7ffd033198a)

**Comment by newville** [comment]
@zhimin-z  Yes, the problem still exists.  Is there any understanding of the cause of the problem?  Has anyone investigated what is going on in the code?  I'm afraid this is a bit outside of my area of expertise.

**Comment by zhimin-z** [comment]
I have no idea...anyone who is in charge of latex functionality in this repo?

**Comment by pfebrer** [comment]
I'm tagging @alexcjohnson since he seems to be the last to have committed something to jupyterlab-plotly other than version changes :)

It looks like this is an easy fix that just needs an update to use the latest MathJax version, or somehow forcing jupyterlab to use an oldest one, I don't know if that's possible.

**Comment by newville** [comment]
@pfebrer @alexcjohnson Thanks for looking into this.  Is there anything the rest of us can do to help further troubleshoot or fix this issue?  I'll admit that I do not have a deep understanding of how plotply.py and plotly.js interact or which version of mathjax  
is actually being used.

**Comment by pfebrer** [comment]
I don't know, everything I know is that jupyterlab uses MathJax `3.2` and plotly's docs use `2.7`, but in principle version `3` should work as well according to https://github.com/plotly/plotly.js/pull/6073

I don't even know how to modify `jupyterlab-plotly` and use a modified version of the extension. I have tried to `pip install -e .` the plotly repo to play a little bit with it but then jupyter doesn't render anything.

**Comment by newville** [comment]
@pfebrer  Thanks, sorry to be dense, but what precisely would one have to do to make it work?   Could that be folded into the plotly-py code?

The issue is kind of 3 months old, and it seems that there has been no explanation or acknowledgment of this issue.  
But, I don't really know what that means. Maybe the solution is obvious to someone?

**Comment by pfebrer** [comment]
I don't know, if I knew I would have submitted a PR already :)

I already tried to fix it and I couldn't even set up a dev environment. I'm not a plotly developer so I don't have more time to spend on it unfortunately.

They already marked it as a bug, so I guess they'll address it whenever they have time. The bug is very annoying and I guess plenty of people will find it since everyone uses jupyter notebooks, so eventually they will have to fix it if they don't want people to move to other packages because of this very small detail.

**Comment by pfebrer** [comment]
> Could that be folded into the plotly-py code?

My impression for this was that `plotlyjs` should be modified (not just the javascript in `jupyterlab-plotly`), but I might be wrong.

**Comment by newville** [comment]
@pfebrer @alexcjohnson it looks like maybe this is related to  #3661, #3696, and #3706, first identified 18 months ago and related to plotly using MathJax 2 when other code (say, Jupyter) has already loaded MathJax 3.  

FWIW, the last release of MathJax2 was 3 years ago. MathJax 3 has been actively developed, and MathJax 4 is now in beta.

**Comment by alexcjohnson** [comment]
As @pfebrer pointed out, Plotly.js has been able to use MathJax v3 or v2 since https://github.com/plotly/plotly.js/pull/6073 which was released in v2.10 (March 2022), so the problem is something about how we're detecting or invoking MathJax in certain environments.

I have to admit my own attempts a few months ago to get multiple versions of Jupyter and Jupyterlab installed side-by-side so I could test and fix these issues left me with a borked environment that wouldn't run anything correctly, and I haven't tried to go back to it since then. So if someone else wants to give this a shot we'd welcome a PR - otherwise we'll get back to it at some point but I can't say when.

@newville thanks for the note about MathJax 4, we'll give it a look! We'll certainly want to get it working in Plotly.js and Dash, regardless of the issues we've had in the various Jupyter environments.

**Comment by newville** [comment]
@alexcjohnson Thanks for the response.    For me, running a standalone plot script, as in https://plotly.com/python/LaTeX/  from a plain Python shell works fine (Python 3.10.12  or 3.11.5 on MacOS) -- it opens a browser in Chrome and the plot correctly renders the LaTeX.

Running that same script in a Jupyter Notebook fails to render the LaTeX.  To reproduce, one should be able to install Miniforge (https://github.com/conda-forge/miniforge), activate that environment, and then install with 

```
mamba install -c conda-forge notebook pandas plotly
```

I might be willing to help, but I am having difficulty even navigating where to start here (not very fluent with the Jupyter code or Plotly).   I cannot really even figure out where the
```
Javascript error: cannot read properties of undefined
```
is coming from.

**Comment by pfebrer** [comment]
>  I cannot really even figure out where the
> 
> Javascript error: cannot read properties of undefined
> is coming from.

If you open the browser console (F12) you'll see some more information. But the JavaScript files are minified so it's really an impossible mission to debug from there.

That's why I tried to set up a dev environment, but I failed. @alexcjohnson maybe if you can point us to somewhere where it is explained how to set up the dev environment for the whole stack (plotly.py, jupyterlab-plotly, plotlyjs), we could try to help. Otherwise, at least in my case, it feels too difficult to try to contribute because I have to figure out everything myself and I don't know if I'm doing things right.

**Comment by newville** [comment]
@alexcjohnson @pfebrer Just in case it wasn't obvious (or in case it helps with debugging), downgrading to jupyter notebook < 7 does avoid this problem:
```
pip install "notebook<7"
```

**Comment by newville** [comment]
@alexcjohnson @pfebrer any thoughts on whether progress is going to be possible on this?  Help (or an attempt to help, at least) was offered, but the very low response rate here is less than encouraging.  I hope all is well.

**Comment by newville** [comment]
@alexcjohnson @pfebrer any updates on this? 

Perhaps a pull request on the documentation (https://github.com/plotly/plotly.py/blob/master/doc/python/LaTeX.md) to change that to state that latex formatting of strings is no longer supported in Jupyter would be in order?  The current statement there is misleading at best.

**Comment by pfebrer** [comment]
I have no idea how to fix it and I'm not part of the plotly team, so I don't see how I can help, sorry :(

I would like to see this fixed as well, but there's nothing I can do.

**Comment by newville** [comment]
@pfebrer Thanks - I am in the same position. It seems like the developers are either ignoring this or won't be able to fix it.  I think that perhaps the best option is to admit that this is not working, and put in a PR to change to documentation to say that latex strings do not work with Jupyter Notebook 7.

**Comment by LunarLanding** [comment]
Currently migrating my report from https://github.com/plotly/plotly.js/issues/5374 to here.
On plotly 4.22, jupyter lab 4.2.3, python 3.11.9, firefox 127.0.2, OSX Sonoma 14.5:
```sh
import plotly.graph_objects as go
f=go.Figure()

f.update_yaxes(title=dict(text="$log_{10}(D)$", font_size=16))

f.show()
```
<img width="1223" alt="image" src="https://github.com/plotly/plotly.py/assets/4441338/6275a606-4060-4832-9a5d-4979c3d3dacb">

**Comment by newville** [comment]
@LunarLanding @timoklein Yeah, the lack of attention and the lack of response to this issue is pretty disappointing. It was raised 10 months ago now.  And as you say, it is a dealbreaker for using Plotly with JupyterLab.   

Both Plotly and JupyterLab explicitly say in multiple places that they work together, and both explicitly say that Latex strings are supported and render correctly.  The fact is -- and has been for a very long time -- that they do not.  Using latex strings with Plotly in JupyterLab or Notebook>7 straight-up breaks any rendering by Plotly.

The Plotly developers continue to not acknowledge this bug and have shown no inclination to address it.

**Comment by marthacryan** [comment]
Hey all! @newville @LunarLanding @timoklein I'm investigating how to fix this, but in the meantime I've had some success with changing the default renderer in `plotly.io` like this (before calling `show`):
```
import plotly.io as pio
pio.renderers.default = 'iframe'
```

**Comment by newville** [comment]
@marthacryan Thanks very much for responding and for the helpful suggestion.  That does work for the couple of examples I've tried.  So that is very promising.

I have not looked at the code in any detail. Is there some documentation or more information about these renderers?  

When using `'iframe'`, a folder called "iframe_figures" is created in the working directory, which has non-trivial HTML files (it appears one per figure, though I am not certain of that).   These seem to not be cleaned up when the notebook is closed.  Would it be possible to either clean these on exit or perhaps write these files to a common temporary folder like /tmp?

FWIW, these HTML files appear to load https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js, which is 5+ years old.
Is that necessary, or could a more recent version be used?

Anyway, that seems very promising. Thanks!

**Comment by archmoj** [comment]
Thanks very much @marthacryan.
It looks like [the default renderer](https://plotly.com/python/renderers/) is not correctly set on `plotly.py`. 
What do you think?

**Comment by pfebrer** [comment]
In other words, if plotly creates extra folders on my filesystem by default it is more probable that I stop using plotly than if latex doesn't work by default :sweat_smile:

**Comment by newville** [comment]
@pfebrer @marthacryan Well, I might be OK with a renderer that creates a temporary file if they are cleaned up or can be centralized to make them easier to clean up (perhaps /tmp or $HOME/.plotly_iframe).   Maybe add a method to clean temporary and/or old temporary plotly-generated files that can be registered to run in an "atexit" handler? 

But, it seems like the observation that the "iframe" renderer is OK must help isolate the problem. For the "plotly_mimetype" renderer, I see "application/vnd.plotly.v1+json", and that is as far as I can go.  That is basically "compiled, indecipherable code that plotly devs own".  If that does not work and the developers are unable to fix the bug they caused, then yeah "don't use" is the only sane response.  

Just to be clear: plotly does not work with Jupyter Notebook > 7.   Workarounds that have unfortunate side effects point to solutions to this problem, but the problem is not "fixed".

For sure, responding to a bug report is appreciated, but taking multiple months to respond is kind of a problem in itself, and now the developers have 2 problems to solve: the technical "how to render plots with Jupyter Notebook" (seems like a pretty basic "fitness of purpose" sort of requirement) and the social "how to convince people who want to use your software and report problems in the forum you set up to believe your software is actually being supported".

**Comment by pfebrer** [comment]
> But, it seems like the observation that the "iframe" renderer is OK must help isolate the problem.

Well, showing it in a separate tab with `fig.show("browser")` has always worked, so the `iframe` solution is a nice discovery to keep things in the notebook but I don't think it helps explain the bug, it is just being shown in an isolated `iframe`.

> For sure, responding to a bug report is appreciated, but taking multiple months to respond is kind of a problem in itself, and now the developers have 2 problems to solve: the technical "how to render plots with Jupyter Notebook" (seems like a pretty basic "fitness of purpose" sort of requirement) and the social "how to convince people who want to use your software and report problems in the forum you set up to believe your software is actually being supported".

I don't pay for using plotly, so all I can do is be grateful that they have open sourced the library and hope that they'll find time to interact with issues and fix the bugs :)

**Comment by marthacryan** [comment]
> I have not looked at the code in any detail. Is there some documentation or more information about these renderers?

Yep, this was already linked in the thread somewhere but the documentation is here: https://plotly.com/python/renderers/

> When using `'iframe'`, a folder called "iframe_figures" is created in the working directory, which has non-trivial HTML files (it appears one per figure, though I am not certain of that). These seem to not be cleaned up when the notebook is closed. Would it be possible to either clean these on exit or perhaps write these files to a common temporary folder like /tmp?

I think rather than changing the location of those files, I'll just try to fix up the default renderer. 

> FWIW, these HTML files appear to load https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js, which is 5+ years old. Is that necessary, or could a more recent version be used?

Yep this is my theory right now! I still need to look into what's going on exactly but it looks to be related to the MathJax version.

> Anyway, that seems very promising. Thanks!

Yay glad it's helping at least in a temporary way!

**Comment by marthacryan** [comment]
> It looks like [the default renderer](https://plotly.com/python/renderers/) is not correctly set on `plotly.py`. What do you think?

@archmoj I think the better solution will be to fix up that default renderer instead of changing to iframe.

**Comment by marthacryan** [comment]
Ah another quick solution here: try `pip install jupyterlab_mathjax2`. It seems to work locally for me, and that way you can avoid the extra files. Again, still trying to fix this default renderer, but hopefully it'll be a good solution while I'm looking into it

**Comment by newville** [comment]
@marthacryan Thanks for the suggestion.  When I do that, then I do not see any text labels at all (for the example at the start of this discussion).  Can you give an example script that works for you?

**Comment by marthacryan** [comment]
@newville Oh interesting. This is what I'm seeing for that example in both JupyterLab and Jupyter Notebook (v7): 
![image](https://github.com/plotly/plotly.py/assets/6673460/2a23c71b-354a-4105-a65f-ec776f6b0a6b)
This is with python 3.12 and 3.8 in a conda env and in python venv. I know calling `jupyter` in the shell sometimes doesn't point to the environment's version of jupyter, do you want to try running `which jupyter` and see if it's pointing to new installation of jupyter in your env?

**Comment by newville** [comment]
@marthacryan Thanks.    
Running `jupyter notebook` with a Python 3.11 environment and with `jupyterlab_mathjax2` installed,  I get this:

<img width="1207" alt="Screenshot 2024-07-09 at 2 06 51 PM" src="https://github.com/plotly/plotly.py/assets/304251/46c9d598-2238-4d76-938c-5f022a0572aa">

**Comment by LunarLanding** [comment]
@marthacryan iframe mode works, thanks! I don't mind the folders much, although it would be better if I could choose the name to something like .iframes.
`jupyterlab_mathjax2` also works.
The ideal solution for me would be for it to just work by loading plotly / mathjax rendering from the internet, and not saving it locally, either in files or in the cell output. Seems the second solution does just that.
For the second method, I also tested that on page reload, and on cold start (jupyter server started, notebook opened without running cells) everything works. 
First a preview saved as png is shown, then both a svg + png are shown, then just the svg without the rendered latex, then the latex renders.

However, I remember that there were other issues that made plotly latex rendering subpar compared to matplotlib.
Namely, mixing math context '$$' and text results in the text being ignored (see screenshot).

@newville I think you forgot a rawstring marker (R'$\alpha...') but I noticed that even adding it, mixing normal text and latex in a string results in the normal text being ignored. But I get some latex rendered, so there is something different between my environment and yours. I think it is because I installed not only the notebook, but jupyter lab, and specifically jupyterlab_mathjax2.

```python
import sys,notebook,plotly
print(sys.version,sys.platform)
for m in (plotly,notebook):
    print(m.__name__,m.__version__)
import plotly.graph_objects as go
f=go.Figure()
f.update_layout(title=dict(text=R'$\alpha$ Title'))
# f.update_yaxes(title=dict(text="a label $log_{10}(D)$", font_size=16))
f.show()
```
<img width="1186" alt="image" src="https://github.com/plotly/plotly.py/assets/4441338/c498d403-7c5d-4033-8109-952e67c611cb">

### Reproducing:

```sh
micromamba create -n plotly_latex_bug jupyterlab plotly jupyterlab_mathjax2
micromamba run -n plotly_latex_bug jupyter lab
```

**Comment by archmoj** [comment]
What is the version of Mathjax on window when using default renderer?

**Comment by newville** [comment]
@marthacryan @LunarLanding The previous screen grab I did was on macOS.  On Windows, with the `jupyterlab-mathjax2` hack installed, I get

![WithNotebook7_mathjax2](https://github.com/plotly/plotly.py/assets/304251/bad3408d-a3bb-4394-ac93-857001874482)

So, the problem is not tied to one system or platform.

**Comment by newville** [comment]
@gvwilson Do I read the P3 label you added as meaning "not pressing"?  

Can you verify that you understand that this problem is actually a show-stopper for using plotly with the latest version of jupyter and jupyterlab?   The first paragraph of the readme advertising this project (https://github.com/plotly/plotly.py/blob/master/README.md) literally says:

```
Inside [Jupyter](https://jupyter.org/install) (installable with pip install "jupyterlab>=3" "ipywidgets>=7.6"):

import plotly.express as px
fig = px.bar(x=["a", "b", "c"], y=[1, 3, 2])
fig.show()
```

but that for anyone wanting to, you know, plot scientific data, that

```
import plotly.express as px
fig = px.bar(x=[r"$\alpha$", r"$\beta$", r"$\gamma$"], y=[1, 3, 2])
fig.show()
```

fails to render a plot.

Please verify that this issue is actually considered "not pressing".

**Comment by gvwilson** [comment]
Hi @newville - "not pressing" means that no-one on my team is actively working on it in the current cycle. If other issues take less time to resolve than expected, we'll bump its priority and find someone to tackle it. PRs from the community are always welcome, and we'll review them as quickly as we can. Thanks - @gvwilson

**Comment by newville** [comment]
@gvwilson Oh.  "not pressing" being short-hand for "not being actively worked on" seems like an odd usage to me.
Is that "If we are working on it, it is urgent, if we are not working on it, it is not urgent"?? 

As with many open-source projects, it is very hard to know from GitHub alone who is on "the team"?  It seems like you view "your team" and "the community" are different things.  That could be self-fulfilling ;)

As it happens, this Issue was first raised almost exactly one year ago.  

It sure *seems* like the Plotly developers will not be able to solve this.  I hope that is not so, as it is unfortunately cripples using Plotly with Jupyter for many people with scientific data.  I wish you all the best, but it is not clear to me who your target audience is.  There are other graphics libraries out there.

**Comment by gdebrun2** [comment]
I don't know why this is closed as the issue persists. I have tried fresh conda/virtual envs, chrome, firefox, and vscode on both windows and Mac. I've also tried notebook v6/v7 and mathjax 2/3 but the default jupyterlab and vscode renderers refuse to render latex. The iframe renderer works fine in jupyterlab but does not work in vscode, so there is no interactive renderer that works within vscode.

**To reproduce:**

1. create a new env, install plotly and jupyterlab
2. Run the following code:
 ```
from plotly.graph_objects import Figure
fig = Figure()
fig.update_layout(title="$Title$")
fig.update_yaxes(title_text=r"$n^{2}$", title_font_size=20)
fig.update_xaxes(title_text="label", title_font_size=20)
```

The generated figure lacks latex labels. In some instances, the labels were added as unrendered raw strings (in vscode and in safari).

<img src=https://github.com/user-attachments/assets/1aa48d28-6a45-441f-8761-a1fe808a1baa>


**Console Output**

```
Refused to execute script from 'http://localhost:8888/static/nbclassic/components/MathJax/jax/output/SVG/fonts/TeX/fontdata.js?V=2.7.9' because its MIME type ('text/html') is not executable, and strict MIME type checking is enabled.
```

Firefox had a similar error. 

**Packages**

<details>

<img src=https://github.com/user-attachments/assets/8d1f1e76-a5fd-4b31-8f58-99771089f653 width=50%>

These packages are what produced the attached figure, but to reiterate I've tried many different versioning combinations.

</details>

**Comment by pfebrer** [comment]
How do you install plotly? There hasn't been a public release yet since the bug was fixed.

**Comment by gdebrun2** [comment]
@pfebrer Ah, my bad I thought this was in  the latest release. But to answer your question, either ```pip install plotly``` or ```conda install -c conda-forge plotly```

**Comment by pfebrer** [comment]
Then I guess it's just a matter of waiting for the next release 👍

**Comment by RafalSkolasinski** [comment]
`6.0.0rc0` fixes it in VS Code but in JupyterLab it still remains broken

**Comment by marthacryan** [comment]
@moonlighthalfwindow Thanks for checking out the rc! This fix was intended to fix the Jupyter Notebook 7 interface and JupyterLab 4 interface. I wasn't aware of the vscode issues with Latex but I can reproduce them and it appears that they've been going on for a long time based on this issue: https://github.com/microsoft/vscode-jupyter/issues/8131. There is a workaround in that issue that I'll copy over here that did fix the behavior for me in vscode:
```
from IPython.display import display, HTML

display(HTML(
    '<script type="text/javascript" async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_SVG"></script>'
))
```
I will work on adding an actual fix to address this but I hope it helps you in the meantime!

**Comment by marthacryan** [comment]
> `6.0.0rc0` fixes it in VS Code but in JupyterLab it still remains broken

@RafalSkolasinski Thanks for trying out the rc! Which version of JupyterLab are you using?

**Comment by RafalSkolasinski** [comment]
> Thanks for trying out the rc! Which version of JupyterLab are you using?

Well, that's embarrassing... I tried again the minimal example from the above
```
from plotly.graph_objects import Figure
fig = Figure()
fig.update_layout(title="$Title$")
fig.update_yaxes(title_text=r"$n^{2}$", title_font_size=20)
fig.update_xaxes(title_text="label", title_font_size=20)
```
and it worked fine...

I did, however, encounter issue when trying to create `fig = go.FigureWidget()` after upgrading plotly to RC
```
ImportError: Please install anywidget to use the FigureWidget class
```
but that's probably subject for a new issue....

**Comment by marthacryan** [comment]
@RafalSkolasinski I'm so glad that it's working for you in JL! And the anywidget error is actually not a bug but the new behavior. You can just `pip install anywidget` and it should go away. This is replacing the `ipywidgets` dependency for plotly.

**Comment by RafalSkolasinski** [comment]
Oh... I really thought anywidget is any widget... Like press a key... Could the error message then somehow include word package to remove the confusion? :)
